File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/NamespaceBlockSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class NamespaceBlockSyntax : DeclarationStatementSyntax
	{
		internal NamespaceStatementSyntax _namespaceStatement;

		internal SyntaxNode _members;

		internal EndBlockStatementSyntax _endNamespaceStatement;

		public NamespaceStatementSyntax NamespaceStatement => GetRedAtZero(ref _namespaceStatement);

		public SyntaxList<StatementSyntax> Members
		{
			get
			{
				SyntaxNode red = GetRed(ref _members, 1);
				return new SyntaxList<StatementSyntax>(red);
			}
		}

		public EndBlockStatementSyntax EndNamespaceStatement => GetRed(ref _endNamespaceStatement, 2);

		internal NamespaceBlockSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal NamespaceBlockSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, NamespaceStatementSyntax namespaceStatement, SyntaxNode members, EndBlockStatementSyntax endNamespaceStatement)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.NamespaceBlockSyntax(kind, errors, annotations, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.NamespaceStatementSyntax)namespaceStatement.Green, members?.Green, (Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax)endNamespaceStatement.Green), null, 0)
		{
		}

		public NamespaceBlockSyntax WithNamespaceStatement(NamespaceStatementSyntax namespaceStatement)
		{
			return Update(namespaceStatement, Members, EndNamespaceStatement);
		}

		public NamespaceBlockSyntax WithMembers(SyntaxList<StatementSyntax> members)
		{
			return Update(NamespaceStatement, members, EndNamespaceStatement);
		}

		public NamespaceBlockSyntax AddMembers(params StatementSyntax[] items)
		{
			return WithMembers(Members.AddRange(items));
		}

		public NamespaceBlockSyntax WithEndNamespaceStatement(EndBlockStatementSyntax endNamespaceStatement)
		{
			return Update(NamespaceStatement, Members, endNamespaceStatement);
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			return i switch
			{
				0 => _namespaceStatement, 
				1 => _members, 
				2 => _endNamespaceStatement, 
				_ => null, 
			};
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			return i switch
			{
				0 => NamespaceStatement, 
				1 => GetRed(ref _members, 1), 
				2 => EndNamespaceStatement, 
				_ => null, 
			};
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitNamespaceBlock(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitNamespaceBlock(this);
		}

		public NamespaceBlockSyntax Update(NamespaceStatementSyntax namespaceStatement, SyntaxList<StatementSyntax> members, EndBlockStatementSyntax endNamespaceStatement)
		{
			if (namespaceStatement != NamespaceStatement || members != Members || endNamespaceStatement != EndNamespaceStatement)
			{
				NamespaceBlockSyntax namespaceBlockSyntax = SyntaxFactory.NamespaceBlock(namespaceStatement, members, endNamespaceStatement);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(namespaceBlockSyntax, annotations);
				}
				return namespaceBlockSyntax;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/NamespaceBlockSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class NamespaceBlockSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.DeclarationStatementSyntax
		Friend _namespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceStatementSyntax

		Friend _members As SyntaxNode

		Friend _endNamespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax

		Public ReadOnly Property EndNamespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax)(Me._endNamespaceStatement, 2)
			End Get
		End Property

		Public ReadOnly Property Members As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)
			Get
				Return New SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)(MyBase.GetRed(Me._members, 1))
			End Get
		End Property

		Public ReadOnly Property NamespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceStatementSyntax
			Get
				Return MyBase.GetRedAtZero(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceStatementSyntax)(Me._namespaceStatement)
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As Microsoft.CodeAnalysis.VisualBasic.SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal namespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceStatementSyntax, ByVal members As SyntaxNode, ByVal endNamespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.NamespaceBlockSyntax(kind, errors, annotations, DirectCast(namespaceStatement.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.NamespaceStatementSyntax), If(members IsNot Nothing, members.Green, Nothing), DirectCast(endNamespaceStatement.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax)), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitNamespaceBlock(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitNamespaceBlock(Me)
		End Sub

		Public Function AddMembers(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax
			Return Me.WithMembers(Me.Members.AddRange(items))
		End Function

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.SyntaxNode
			Dim syntaxNode As Microsoft.CodeAnalysis.SyntaxNode
			Select Case i
				Case 0
					syntaxNode = Me._namespaceStatement
					Exit Select
				Case 1
					syntaxNode = Me._members
					Exit Select
				Case 2
					syntaxNode = Me._endNamespaceStatement
					Exit Select
				Case Else
					syntaxNode = Nothing
					Exit Select
			End Select
			Return syntaxNode
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Dim namespaceStatement As SyntaxNode
			Select Case i
				Case 0
					namespaceStatement = Me.NamespaceStatement
					Exit Select
				Case 1
					namespaceStatement = MyBase.GetRed(Me._members, 1)
					Exit Select
				Case 2
					namespaceStatement = Me.EndNamespaceStatement
					Exit Select
				Case Else
					namespaceStatement = Nothing
					Exit Select
			End Select
			Return namespaceStatement
		End Function

		Public Function Update(ByVal namespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceStatementSyntax, ByVal members As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax), ByVal endNamespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax
			Dim namespaceBlockSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax
			If (namespaceStatement <> Me.NamespaceStatement OrElse members <> Me.Members OrElse endNamespaceStatement <> Me.EndNamespaceStatement) Then
				Dim namespaceBlockSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.NamespaceBlock(namespaceStatement, members, endNamespaceStatement)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				namespaceBlockSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, namespaceBlockSyntax1, namespaceBlockSyntax1.WithAnnotations(annotations))
			Else
				namespaceBlockSyntax = Me
			End If
			Return namespaceBlockSyntax
		End Function

		Public Function WithEndNamespaceStatement(ByVal endNamespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.EndBlockStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax
			Return Me.Update(Me.NamespaceStatement, Me.Members, endNamespaceStatement)
		End Function

		Public Function WithMembers(ByVal members As SyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax
			Return Me.Update(Me.NamespaceStatement, members, Me.EndNamespaceStatement)
		End Function

		Public Function WithNamespaceStatement(ByVal namespaceStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceStatementSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.NamespaceBlockSyntax
			Return Me.Update(namespaceStatement, Me.Members, Me.EndNamespaceStatement)
		End Function
	End Class
End Namespace